workspaces/ui-v2/src/pages/docs/components/ShapeEditor.tsx

```diff
@@ -0,0 +1,464 @@
+import React, { FC, useMemo } from 'react';
+import {
+  makeStyles,
+  lighten,
+  IconButton,
+  ButtonGroup,
+  Button,
+  TextField,
+} from '@material-ui/core';
+import { Check as CheckIcon } from '@material-ui/icons';
+import ClassNames from 'classnames';
+import Color from 'color';
+import DeleteOutlineIcon from '@material-ui/icons/DeleteOutline';
+
+import { IFieldDetails, IShapeRenderer } from '<src>/types';
+import { JsonType } from '@useoptic/optic-domain';
+import * as Theme from '<src>/styles/theme';
+
+export const ShapeEditor: FC<{
+  fields: IFieldDetails[];
+  selectedFieldId: string | null;
+  setSelectedField: (fieldId: string | null) => void;
+}> = ({ fields, selectedFieldId, setSelectedField }) => {
+  const classes = useStyles();
+
+  let onChangeFieldDescription = useMemo(
+    () => (fieldId: string, description: string) => {
+      console.log('description changed for field', fieldId, description);
+    },
+    []
+  );
+
+  return (
+    <div className={classes.container}>
+      {fields.length > 0 && (
+        <ul className={classes.rowsList}>
+          {fields.map((field) => (
+            <li className={classes.rowListItem}>
+              <Row
+                field={field}
+                description={''} // TODO FLEB: wire me in
+                selected={selectedFieldId === field.fieldId}
+                onChangeDescription={onChangeFieldDescription}
+                onSelect={setSelectedField}
+              />
+            </li>
+          ))}
+        </ul>
+      )}
+    </div>
+  );
+};
+
+const Row: FC<{
+  field: IFieldDetails;
+  description: string;
+  selected: boolean;
+
+  onChangeDescription?: (fieldId: string, description: string) => void;
+  onSelect?: (fieldId: string) => void;
+}> = function ShapeEditorRow({
+  field,
+  description,
+  selected,
+  onChangeDescription,
+  onSelect,
+}) {
+  const classes = useStyles();
+
+  const onClickFieldHeader = useMemo(
+    () => () => {
+      if (onSelect) onSelect(field.fieldId);
+    },
+    [onSelect, field.fieldId]
+  );
+
+  const onChangeType = useMemo(
+    () => (type: JsonType, enabled: boolean) => {
+      console.log('changed type for field', type, enabled, field.fieldId);
+    },
+    [field.fieldId]
+  );
```

Comment:
This function signature might need to change, but I'll tackle this when i hook this up into redux actions that I've set up.

Reply:
If it does, then it does. Most important to me is that all of this stuff bubbles up to `ShapeEditor` (and preferable in a way that makes sense in isolation), and have all child components be "unconnected" in oldskool Redux parlance.  You end up handing down props a bit more, but it makes it so much easier to evolve and re-use parts if you don't have to worry about all that context. Preferably, even `ShapeEditor` is unaware of redux or context, and could be hooked up to other state solutions (or mocks) just as easily.

Reply:
I'd agree with this, however if we do that and we change the field, that will result in all fields rerendering on a field change (if we pass down top level props) - so in this case it might be better to have access to the redux store. The biggest point of confusion is if we reuse this component across pages; needing to be renderered in a redux store can get messy unless the assumption is that there is _always_ a redux store.

There's definitely other ways to handle this, but it is probably easier to access props via redux selectors passing down props + handling render trigger behavior

Reply:
Sure, we can definite do performance optimization in a bunch of different ways - and the fact that we already have functions + children that already cause re-renders does not mean we shouldn't do this. The ideal is that the component is portable with a clear split between them, and you can definitely keep that interface while optimizing performance, but that comes with a higher cost than using redux selectors (which have this logic built in). 

Context has advantages (in this example, redux) over passing props with render optimization.

> If that were to become a performance problem (which it really won't)

We have had performance issues with this page in the past where a change in a large shape caused a large amount of nodes to be rerendered - causing noticeable UI jank - so I wouldn't say it's something that will never happen and we never need to worry about this.